src/network/resolver.rs

```diff
@@ -0,0 +1,89 @@
+use std::{
+    collections::HashMap,
+    net::Ipv4Addr,
+    sync::{Arc, Mutex},
+    thread::{Builder, JoinHandle},
+};
+use tokio::{
+    runtime::Runtime,
+    sync::mpsc::{self, Sender},
+};
+use trust_dns_resolver::AsyncResolver;
+
+pub type Result<T> = std::result::Result<T, failure::Error>;
+pub type IpTable = Arc<Mutex<HashMap<Ipv4Addr, String>>>;
```

Comment:
I like the IpTable type! I'd be happy if we could also be using it in main.rs (instead of the `HashTable::new` when we create `ip_to_host`). Could we remove the `Arc<Mutex<>>` wrapper though? I don't like accepting and expecting them, but rather handling them in one place and having their "users" not be aware of them unless they have to be (eg. with Mutex sometimes). Does that make sense?

Reply:
Sure, I will use the type in main.rs too! Without the `Arc<Mutex<>>` as you suggest. But the resolver will have to take `Arc<Mutex<IpTable>>` as it will be modifying the map from another thread, tell me if I missed something about your comment :)

Reply:
I'll have to see the implementation, but there might be a creative way to invert stuff around to make it work (you can look at what we did with the previous dns_queue for inspiration). If it doesn't work out, no sweat.